docs/instrumentation.md

```diff
@@ -0,0 +1,283 @@
+## TPM driver instrumentation
+
+For development purposes, a set of scripts was created that allows for the
+instrumentation of a specific module, allowing for tracing the execution
+of module functions.
+
+This could be especially useful in the process of development of different
+protocol backends in our `lpntpn` like LPC or I2C.
+
+For now, scripts are located inside our fork of
+[ms-tpm-20-ref](https://github.com/lpn-plant/ms-tpm-20-ref) repository.
+Look for `Samples/Nucleo-TPM/scripts/kfunc_tracer` directory on the dev branch.
+
+We use the integrated functionality that gcc provides '-finstrument-function'
+for function instrumentation.
+This step allows us to register exact function call addresses and log them to
+dmesg.
+
+Next, we match the function addresses against kernel processes address space
+that's located in `/proc/modules` to identify loaded module, that fn call
+belongs to. By subtracting function address with start memory location of loaded
+module, we obtain the real offset of a function call, that could be easily
+matched with an exact function name.
+
+The last step is to extract functions memory offsets located in non-striped
+kernel modules. For this purpose we are using a simple readelf call:
+`readelf -s MODULE_NAME.ko | grep FUNC`.
+
+This process requires kernel recompilation, execution of module startup and
+parsing of resulting trace files.
+
+### Prepare Raspberry Pi
+
+We are using Raspberry Pi 3, but with minor modifications, this process could
+be used on different targets.
+
+**Flash SD card with raspbian lite os.**
+
+At the moment of writing image `2021-05-07-raspios-buster-armhf-lite.img` was
+used.
+
+**Enable ssh.**
+
+Execute on raspberry pi with connected keyboard and display
+
+```shell
+# upgrade
+sudo apt update
+sudo apt upgrade
+
+# change default password
+passwd
+
+# enable ssh
+sudo systemctl enable ssh
+sudo systemctl start ssh
+```
+
+**Check ssh connection.**
+
+Environmental variable `RASPI_IP` will be used several more times during this
+guide.
+
+Run on your host machine to test ssh connection.
+```shell
+export RASPI_IP=pi@192.168.8.170
+ssh-copy-id -i ~/.ssh/id_rsa.pub $RASPI_IP
+ssh $RASPI_IP echo "success"
+# should echo success without prompting user to enter password
+```
+
+### Kernel compinaltion
+
+**Prepare build environment.**
+
+Ubuntu 20.04 LTS was used in this example.
+
+Install dependencies and get source code.
+```shell
+sudo apt install crossbuild-essential-armhf     # default toolchain
+sudo apt install python-is-python3              # no python2 in ubuntu 20 fix
+sudo apt install bison flex
+sudo apt install qtbase5-dev pkg-config         # xconfig
+sudo apt install libncurses5-dev                # menuconfig
+
+cd git
+git clone –-depth=1 https://github.com/raspberrypi/linux
```

Comment:
@acwiek This will use main branch, which is a moving target and will be different depending on the time of use. 
It would be nice to use the specific commit which was used in your case.